packages/stateManager/index.ts

```diff
@@ -1,104 +1,636 @@
+  addAcceptedCurrencies: ShopCurrencies[];
+  addPayee: {
+    addr: `0x${string}`;
+    callAsContract: boolean;
+    chainId: number;
+    name: string;
+  } | null;
+  publishedTagId: `0x${string}`;
+  profilePictureUrl: string;
+};
+type ShopObjectTypes = Item | Tag | KeyCards | Order | ShopManifest;
 
-type ShopObjectTypes = Item | Tag | KeyCards;
-
-// This is a an interface that is used to retrieve and store objects from a persistant layer
+// This is an interface that is used to retrieve and store objects from a persistant layer
 type Store<T extends ShopObjectTypes> = {
   put(key: string, value: T): Promise<void>;
   get(key: string): Promise<T>;
+  iterator(): AsyncIterator<T>;
 };
 
 abstract class PublicObjectManager<
   T extends ShopObjectTypes,
 > extends EventEmitter {
   constructor(
-    protected db: Store<T>,
+    protected store: Store<T>,
     protected client: RelayClient,
   ) {
     super();
   }
-  abstract _processEvent(event: schema.ShopEvents): Promise<boolean>;
-  abstract create(obj: T): Promise<[any, any]>;
-  abstract update(obj: T): Promise<[any, any]>;
-  get(id: string): Promise<T> {
-    return this.db.get(id);
+  abstract _processEvent(event: schema.ShopEvents): Promise<void>;
+  abstract get(key?: string): Promise<T>;
+  get iterator() {
+    return this.store.iterator();
   }
 }
+//We should always make sure the network call is successful before updating the store with store.put
+class ListingManager extends PublicObjectManager<Item> {
+  constructor(store: Store<Item>, client: RelayClient) {
+    super(store, client);
+  }
+  // Process all events for listings.
+  // Convert bytes to hex and save item object to listings store.
+  async _processEvent(event: schema.ShopEvents): Promise<void> {
+    if (event.createItem) {
+      const ci = event.createItem;
+      const id = bytesToHex(ci.eventId);
+      const item = {
+        id,
+        price: ci.price,
+        metadata: bufferToJSON(ci.metadata),
+        tags: [],
+        quantity: 0,
+      };
+      await this.store.put(id, item);
+      this.emit("createItem", item);
+      return;
+    } else if (event.updateItem) {
+      const ui = event.updateItem;
+      const id = bytesToHex(ui.itemId);
+      const item = await this.store.get(id);
+      if (ui.metadata) {
+        item.metadata = bufferToJSON(ui.metadata);
+      }
+      if (ui.price) {
+        item.price = ui.price;
+      }
+      await this.store.put(id, item);
+      this.emit("updateItem", item);
+      return;
+    } else if (event.changeStock) {
+      const cs = event.changeStock;
+      if (cs.itemIds) {
+        await Promise.all(
+          cs.itemIds.map(async (id: Uint8Array, i: number) => {
+            const itemId = bytesToHex(id);
+            const item = await this.store.get(itemId);
+            const diff = cs.diffs ? cs.diffs[i] : 0;
+            item.quantity = item.quantity + diff;
+            await this.store.put(itemId, item);
```

Comment:
```suggestion
            return this.store.put(itemId, item);
```

we need to return the promise here so that it ends up in the list for `Promise.all`.